sonar-python-plugin/src/main/java/org/sonar/plugins/python/SymbolVisitor.java

```diff
@@ -0,0 +1,81 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.python;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import org.sonar.api.batch.sensor.symbol.NewSymbol;
+import org.sonar.api.batch.sensor.symbol.NewSymbolTable;
+import org.sonar.python.api.tree.ClassDef;
+import org.sonar.python.api.tree.FileInput;
+import org.sonar.python.api.tree.FunctionDef;
+import org.sonar.python.api.tree.LambdaExpression;
+import org.sonar.python.semantic.Symbol;
+import org.sonar.python.semantic.Usage;
+import org.sonar.python.tree.BaseTreeVisitor;
+
+public class SymbolVisitor extends BaseTreeVisitor {
+
+  private final NewSymbolTable newSymbolTable;
+
+  public SymbolVisitor(NewSymbolTable newSymbolTable) {
+    this.newSymbolTable = newSymbolTable;
+  }
+
+  @Override
+  public void visitClassDef(ClassDef classDef) {
+    classDef.classFields().forEach(this::handleSymbol);
+    classDef.instanceFields().forEach(this::handleSymbol);
+    super.visitClassDef(classDef);
+  }
+
+  @Override
+  public void visitFunctionDef(FunctionDef functionDef) {
+    functionDef.localVariables().forEach(this::handleSymbol);
+    super.visitFunctionDef(functionDef);
+  }
+
+  @Override
+  public void visitLambda(LambdaExpression lambdaExpression) {
+    lambdaExpression.localVariables().forEach(this::handleSymbol);
+    super.visitLambda(lambdaExpression);
+  }
+
+  @Override
+  public void visitFileInput(FileInput fileInput) {
+    fileInput.moduleFields().forEach(this::handleSymbol);
+    super.visitFileInput(fileInput);
+    newSymbolTable.save();
+  }
+
+  private void handleSymbol(Symbol symbol) {
+    List<Usage> usages = new ArrayList<>(symbol.usages());
+    usages.sort(Comparator.comparingInt(u -> u.tree().firstToken().line()));
+    Usage firstUsage = usages.get(0);
+    NewSymbol newSymbol = newSymbolTable.newSymbol(firstUsage.tree().firstToken().line(), firstUsage.tree().firstToken().column(),
+      firstUsage.tree().lastToken().line(), firstUsage.tree().lastToken().column() + firstUsage.tree().lastToken().value().length());
+    for (int i = 1; i < usages.size(); i++) {
+      Usage usage = usages.get(i);
+      newSymbol.newReference(usage.tree().firstToken().line(), usage.tree().firstToken().column(),
+        usage.tree().lastToken().line(), usage.tree().lastToken().column() + firstUsage.tree().lastToken().value().length());
```

Comment:
Maybe we can store usage.tree() in a dedicated var instead of calling it 4 times ? which makes me realize that we might have a copy/paste bug here : there is a firstUsage.tree() call !!!  Why is this not caught by unit test  ? (seems that ending of reference is not tested)